prototypes/palloc/pool_alloc.c

```diff
@@ -0,0 +1,212 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <unistd.h>
+#include "queue.h"
+#include "pool_alloc.h"
+
+#define align_to_binary_pow(x, a) ((((x) >> (a))+1) << (a))
+
+#define DEBUG 1
+
+#if defined(DEBUG) && DEBUG > 0
+    #define DEBUG_PRINT(text, args ...) fprintf(stderr, "DEBUG: %s:%d:%s(): " text, \
+        __FILE__, __LINE__, __func__, ##args)
+#else
+    #define DEBUG_PRINT(text, args ...)
+#endif
+
+#define panic(text, args ...)                                                 \
+  __extension__({                                               \
+    fprintf(stderr, "FATAL ERROR: %s:%d:%s(): " text, __FILE__, __LINE__, __func__, ##args);     \
+    exit(EXIT_FAILURE);                                         \
+  })
+  
+int page_size;
+
+typedef struct pool_item {
+    uint64_t pi_guard_number; //0xcafebabecafebabe by default, normally isn't changed, so if it has another value, memory corruption has taken place
```

Comment:
When picking magic guard values, it is a good practice to pick one that does not contain any (long) substring more than once. The one you picked is particularly unfortunate: `cafebabe` repeats twice at a word-boundary! So if one happens to write this value to address `X` and afterwards at `X+4`, they overlay perfectly, and it may seem that `X` still holds a valid guard number, even though it was overwritten.